examples/tts/aligner_g2p.py

```diff
@@ -0,0 +1,340 @@
+    audio_len = torch.tensor(audio_data.shape[0], device=device).long().unsqueeze(0)
+
+    # Generate spectrogram
+    spec, spec_len = aligner.preprocessor(input_signal=audio, length=audio_len)
+
+    return spec, spec_len
+
+
+def get_mean_distance_for_word(l2_dists, durs, start_token, num_tokens):
+    """Calculates the mean distance between text and audio embeddings given a range of text tokens.
+
+    This is the same function as in the Aligner Inference tutorial notebook.
+    """
+    # Need to calculate which audio frame we start on by summing all durations up to the start token's duration
+    start_frame = torch.sum(durs[:start_token]).data
+
+    total_frames = 0
+    dist_sum = 0
+
+    # Loop through each text token
+    for token_ind in range(start_token, start_token + num_tokens):
+        # Loop through each frame for the given text token
+        for frame_ind in range(start_frame, start_frame + durs[token_ind]):
+            # Recall that the L2 distance matrix is shape [spec_len, text_len]
+            dist_sum += l2_dists[frame_ind, token_ind]
+
+        # Update total frames so far & the starting frame for the next token
+        total_frames += durs[token_ind]
+        start_frame += durs[token_ind]
+
+    return dist_sum / total_frames
+
+
+def disambiguate_candidates(aligner, text, spec, spec_len, confidence, device, heteronyms, log_file=None):
+    """Retrieves and disambiguate all candidate sentences for disambiguation of a given some text.
+
+    Assumes that the max number of candidates per word is a reasonable batch size.
+
+    Note: This could be sped up if multiple words' candidates were batched, but this is conceptually easier.
+    """
+    # Grab original G2P result
+    aligner_g2p = aligner.tokenizer.g2p
+    base_g2p = aligner_g2p(text)
+
+    # Tokenize text
+    words = [word for word, _ in aligner_g2p.word_tokenize_func(text)]
+
+    ### Loop Through Words ###
+    result_g2p = []
+    word_start_idx = 0
+
+    has_heteronym = False
+
+    for word in words:
+        # Retrieve the length of the word in the default G2P conversion
+        g2p_default_len = len(aligner_g2p(word))
+
+        # Check if word needs to be disambiguated
+        if word in heteronyms:
+            has_heteronym = True
+
+            # Add candidate for each ambiguous pronunciation
+            word_candidates = []
+            candidate_prons_and_lengths = []
+
+            for pron in aligner_g2p.phoneme_dict[word]:
+                # Replace graphemes in the base G2P result with the current variant
+                candidate = base_g2p[:word_start_idx] + pron + base_g2p[word_start_idx + g2p_default_len :]
+                candidate_tokens = aligner.tokenizer.encode_from_g2p(candidate)
+
+                word_candidates.append(candidate_tokens)
+                candidate_prons_and_lengths.append((pron, len(pron)))
+
+            ### Inference ###
```

Comment:
is there a way to add all this to the model's infer() method?

Reply:
Discussed over Slack. TL;DR: The Aligner model doesn't have disambiguation as a main use case, so we are keeping this as a separate script functionality.